schema/dmt/build.go

```diff
@@ -73,3 +35,133 @@ func Compile(ts *schema.TypeSystem, node *Schema) error {
+	if b == nil {
+		return false
+	}
+	return *b
+}
+
+func todoAnonTypeName(nameOrDefn TypeNameOrInlineDefn) string {
+	if nameOrDefn.TypeName != nil {
+		return *nameOrDefn.TypeName
+	}
+	defn := *nameOrDefn.InlineDefn
+	switch {
+	case defn.TypeDefnMap != nil:
+		defn := defn.TypeDefnMap
+		return fmt.Sprintf("Map__%s__%s", defn.KeyType, todoAnonTypeName(defn.ValueType))
+	case defn.TypeDefnList != nil:
+		defn := defn.TypeDefnList
+		return fmt.Sprintf("List__%s", todoAnonTypeName(defn.ValueType))
+	default:
+		panic(fmt.Errorf("%#v", defn))
+	}
+}
+
+func spawnType(ts *schema.TypeSystem, name schema.TypeName, defn TypeDefn) (schema.Type, error) {
+	switch {
+	// Scalar types without parameters.
+	case defn.TypeDefnBool != nil:
+		return schema.SpawnBool(name), nil
+	case defn.TypeDefnString != nil:
+		return schema.SpawnString(name), nil
+	case defn.TypeDefnBytes != nil:
+		return schema.SpawnBytes(name), nil
+	case defn.TypeDefnInt != nil:
+		return schema.SpawnInt(name), nil
+	case defn.TypeDefnFloat != nil:
+		return schema.SpawnFloat(name), nil
+
+	case defn.TypeDefnList != nil:
+		typ := defn.TypeDefnList
+		if typ.ValueType.InlineDefn != nil {
+			return nil, fmt.Errorf("TODO: support anonymous types in schema package")
+		}
+		switch {
+		case typ.Representation == nil ||
+			typ.Representation.ListRepresentation_List != nil:
+			// default behavior
+		default:
+			return nil, fmt.Errorf("TODO: support non-default map repr in schema package")
+		}
+		return schema.SpawnList(name,
+			schema.TypeName(*typ.ValueType.TypeName),
+			todoFromImplicitlyFalseBool(typ.ValueNullable),
+		), nil
+	case defn.TypeDefnMap != nil:
+		typ := defn.TypeDefnMap
+		if typ.ValueType.InlineDefn != nil {
+			return nil, fmt.Errorf("TODO: support anonymous types in schema package")
+		}
+		switch {
+		case typ.Representation == nil ||
+			typ.Representation.MapRepresentation_Map != nil:
+			// default behavior
+		default:
+			return nil, fmt.Errorf("TODO: support non-default map repr in schema package")
+		}
+		return schema.SpawnMap(name,
+			schema.TypeName(typ.KeyType),
+			schema.TypeName(*typ.ValueType.TypeName),
+			todoFromImplicitlyFalseBool(typ.ValueNullable),
+		), nil
+	case defn.TypeDefnStruct != nil:
+		typ := defn.TypeDefnStruct
+		var fields []schema.StructField
+		for _, fname := range typ.Fields.Keys {
+			field := typ.Fields.Values[fname]
+			var typeName schema.TypeName
+			if field.Type.TypeName != nil {
+				typeName = schema.TypeName(*field.Type.TypeName)
+			} else if tname := todoAnonTypeName(field.Type); ts.TypeByName(tname) == nil {
+				typeName = schema.TypeName(tname)
+				// Note that TypeDefn and InlineDefn aren't the same enum.
+				anonDefn := TypeDefn{
+					TypeDefnMap:  field.Type.InlineDefn.TypeDefnMap,
+					TypeDefnList: field.Type.InlineDefn.TypeDefnList,
+					TypeDefnLink: field.Type.InlineDefn.TypeDefnLink,
+				}
+				anonType, err := spawnType(ts, typeName, anonDefn)
```

Comment:
Okay, I had to read this a couple times before I figured out what was up.

Wait, no.  One more time.

... okay, it makes sense now.  It's doing a sort of a, uh, reverse filter.  So code can be reused.  Alright, checks out.

Reply:
Yeah this is awkward as heck, but I also didn't bother with properly documenting it because it's a temporary hack until the schema package supports inline type defns.